### DISCUSSION TITLE
SocketCAN/vcan support in Ubuntu VM

### DISCUSSION THREAD

**Discussion by appelgriebsch** [abiosoft/colima#1108] [Q&A]
Hey, following the discussion here https://github.com/docker/for-mac/issues/4405 I'm looking for a way to use SocketCAN within my colima VM, but it seems the required kernel modules are not available as well....

Can I use a custom-tailored VM instead, or are there any plans to introduce them into the default image?

**Comment by abiosoft** [comment]
> Can I use a custom-tailored VM instead

Not yet, except you build Colima from source. Ability to specify custom images is being considered.

> or are there any plans to introduce them into the default image?

No objections actually, I only need the instructions on how to do that. 
And feel free to contribute if you have the skillset for that, the Ubuntu image is being customized in this script https://github.com/abiosoft/colima-core/blob/main/scripts/qcow.sh

**Comment by abiosoft** [comment]
> Is it reproducible: Yes. Take an Alpine/Debian/Ubuntu docker image, install can-utils (with a package manager or by compiling from [source](https://github.com/linux-can/can-utils.git))

If that's all required, you can simply ssh into the Colima VM and install that.

```sh
# install can-utils
colima ssh -- sh -c "sudo apt update && sudo apt install can-utils"

# restart (if required)
colima restart
```

If that works for you as desired, then I can include the package in the default image going forward.

**Reply by appelgriebsch** [reply]
Sadly it's not. It requires availability of certain kernel modules (namely `vcan`) to be able to create virtual vcan network interfaces. Just checked the build script. As it's using the ubuntu-minimal cloud image I can imagine those kernel modules are not configured at all.... which either means: using a different (larger) base image, or compiling certain kernel modules on your own.... (which makes updating to newer images quite complex)

**Reply by appelgriebsch** [reply]
According to the ticket in docker-for-mac the following minimum set on kernel modules need to be configured

```

#
# CAN Bus support
#

CONFIG_CAN=m
CONFIG_CAN_RAW=m
CONFIG_CAN_BCM=m
CONFIG_CAN_GW=m

#
# CAN Device Drivers
#
CONFIG_CAN_VCAN=m
CONFIG_CAN_DEV=m

```

**Reply by KaSroka** [reply]
Hello, any updates on this? I also need to setup vcan interface from a docker container.

**Reply by nickelsmore-lime** [reply]
I've actually managed to solve this by using a provision script to build and load the module. Perhaps it can just be included in the default kernel in the future?
```
provision:
  - mode: system
    script: |
      #!/bin/bash
      set -eu

      KERNEL_VER=$(uname -r)
      MODULE_DEST="/lib/modules/${KERNEL_VER}/kernel/net/can/vcan.ko.zst"

      # Fast path: module already loadable (built previously or included in kernel)
      if modprobe vcan 2>/dev/null; then
          exit 0
      fi

      # Fast path: module file exists but depmod needs refresh
      if [ -f "$MODULE_DEST" ]; then
          depmod -a
          modprobe vcan
          exit 0
      fi

      echo "vcan: module not found — building from kernel source..."

      SOURCE_VER=$(echo "$KERNEL_VER" | grep -oP '^\d+\.\d+\.\d+')

      apt-get update -qq
      apt-get install -y -qq build-essential \
          linux-headers-${KERNEL_VER} \
          linux-source-${SOURCE_VER} \
          zstd >/dev/null

      WORK_DIR=$(mktemp -d)

      # Extract only the CAN driver source
      tar xf /usr/src/linux-source-${SOURCE_VER}.tar.* \
          -C "$WORK_DIR" --strip-components=1 \
          linux-source-${SOURCE_VER}/drivers/net/can

      # Build against installed kernel headers (provides Module.symvers
      # for correct symbol versioning)
      make -C /lib/modules/${KERNEL_VER}/build \
          M="${WORK_DIR}/drivers/net/can" \
          modules

      # Install the built modules — replace both vcan and can-dev to
      # avoid symbol version mismatches with the pre-installed modules
      CAN_DIR="/lib/modules/${KERNEL_VER}/kernel/net/can"

      # Unload any existing CAN modules before replacing
      rmmod can_gw can_raw can_isotp can_bcm can 2>/dev/null || true

      for ko in "$WORK_DIR"/drivers/net/can/*.ko; do
          zstd -f "$ko" -o "${CAN_DIR}/$(basename "$ko").zst"
      done
      mkdir -p "${CAN_DIR}/dev"
      for ko in "$WORK_DIR"/drivers/net/can/dev/*.ko; do
          zstd -f "$ko" -o "${CAN_DIR}/dev/$(basename "$ko").zst"
      done

      depmod -a
      rm -rf "$WORK_DIR"

      modprobe vcan

      # Auto-load on subsequent boots so the build is never needed again
      echo "vcan" > /etc/modules-load.d/vcan.conf

      echo "vcan: module built and installed successfully."
```

**Comment by jimpnet** [comment]
This worked for me by just doing this:

colima ssh

sudo apt update;sudo apt install linux-modules-extra-$(uname -r)

exit


That gets it working for Docker containers running in Colima.


PLATFORM SCIENCE <https://www.platformscience.com/>

Jim McClure | Distinguished Engineer

***@***.***

4400 Baker Rd, Minnetonka, MN 55343

**Reply by nickelsmore-lime** [reply]
Yeah thats way better. I was overcomplicating it, thanks for that!
